### PR TITLE
feat (worker, api): upload pipeline build artifact with swift temp url

### PR DESCRIPTION
**Pull request by fsamin**
This contains:
* new vendored switf client
* upload artifacts with temp url (for pipeline build only)
* few I/O optimizations
 
To enable it, just set objectstore mode to swift and be careful with auth url

In other PR: next steps will be:

* download pipeline build artifact with temp url
* upload and download workflow artifact with temp url
* download plugin with temp url
* remove old code

### CODE REVIEW

File: engine/api/objectstore/swift.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ovh/cds/engine/api/sessionstore"
+
+	"github.com/ncw/swift"
+	"github.com/ovh/cds/sdk"
```
Comment (bnjjj) [inline engine/api/objectstore/swift.go]: bad news :(

File: engine/api/artifact.go [hunk 2]
```diff
@@ -264,13 +280,155 @@ func (api *API) downloadArtifactDirectHandler() Handler {
 			return sdk.WrapError(err, "downloadArtifactDirectHandler> Could not load artifact with hash %s", hash)
 		}
 
+		log.Debug("downloadArtifactDirectHandler: Serving %+v", art)
+		f, err := objectstore.FetchArtifact(art)
+		if err != nil {
+			return sdk.WrapError(err, "downloadArtifactDirectHandler> Cannot fetch artifact")
+		}
+
```
Comment (sguiheux) [inline engine/api/artifact.go]: defer f.close ?

File: engine/api/objectstore/swift.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
+package objectstore
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ovh/cds/engine/api/sessionstore"
```
Comment (sguiheux) [inline engine/api/objectstore/swift.go]: wrong place

File: engine/api/objectstore/swift.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+	"github.com/ncw/swift"
```
Comment (sguiheux) [inline engine/api/objectstore/swift.go]: wrong place

File: engine/api/objectstore/swift.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	file, errC := s.ObjectCreate(container, object, false, "", "application/octet-stream", nil)
+	if errC != nil {
+		return "", sdk.WrapError(errC, "SwiftStore> Unable to create object %s", object)
+	}
+
+	log.Debug("SwiftStore> copy object %s/%s", container, object)
+	if _, err := io.Copy(file, data); err != nil {
+		return "", sdk.WrapError(err, "SwiftStore> Unable to copy object buffer %s", object)
+	}
+
```
Comment (sguiheux) [inline engine/api/objectstore/swift.go]: defer file.Close + data.close

File: engine/api/plugin.go [hunk 6]
```diff
@@ -256,8 +256,12 @@ func (api *API) downloadPluginHandler() Handler {
 		w.Header().Add("Content-Type", "application/octet-stream")
 		w.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", name))
 
-		if err := objectstore.StreamFile(w, f); err != nil {
-			return sdk.WrapError(err, "downloadPluginHandler> Error while streaming plugin %s", name)
+		if _, err := io.Copy(w, f); err != nil {
+			return sdk.WrapError(err, "downloadPluginHandler> Cannot stream artifact")
+		}
+
```
Comment (sguiheux) [inline engine/api/plugin.go]: close

File: engine/api/workflow_run.go [hunk 7]
```diff
@@ -540,9 +541,17 @@ func (api *API) downloadworkflowArtifactDirectHandler() Handler {
...
 		w.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", art.Name))
 
-		log.Debug("downloadworkflowArtifactDirectHandler: Serving %+v", art)
-		if err := artifact.StreamFile(w, art); err != nil {
-			return sdk.WrapError(err, "downloadworkflowArtifactDirectHandler: Cannot stream artifact")
+		f, err := objectstore.FetchArtifact(art)
+		if err != nil {
+			return sdk.WrapError(err, "downloadArtifactDirectHandler> Cannot fetch artifact")
+		}
+
```
Comment (sguiheux) [inline engine/api/workflow_run.go]: defer f.close

File: engine/api/workflow_run.go [hunk 8]
```diff
@@ -596,8 +605,17 @@ func (api *API) getDownloadArtifactHandler() Handler {
 		w.Header().Add("Content-Type", "application/octet-stream")
 		w.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", art.Name))
 
-		if err := artifact.StreamFile(w, art); err != nil {
-			return sdk.WrapError(err, "Cannot stream artifact %s", art.Name)
+		f, err := objectstore.FetchArtifact(art)
+		if err != nil {
+			return sdk.WrapError(err, "getDownloadArtifactHandler> Cannot fetch artifact")
+		}
+
```
Comment (sguiheux) [inline engine/api/workflow_run.go]: defer f.close

File: engine/api/artifact.go (hunk 2 above)
Comment (fsamin) [reply engine/api/artifact.go]: Nope i need to catch de f.close error

File: engine/api/objectstore/swift.go (hunk 5 above)
Comment (fsamin) [reply engine/api/objectstore/swift.go]: not in defer, i want to catch 'em all

File: engine/api/workflow_run.go (hunk 8 above)
Comment (fsamin) [reply engine/api/workflow_run.go]: nop i want to catch the error

File: engine/api/workflow_run.go (hunk 7 above)
Comment (fsamin) [reply engine/api/workflow_run.go]: not here

**Pull request merged**